File: TellDontAsk/AlarmsProcedural/AlarmServices/TimedAlarmService.cs
using System;
using System.Linq;
using AlarmsProcedural.Adapters;
using AlarmsProcedural.Enums;

namespace AlarmsProcedural.AlarmServices
{
  public static class TimedAlarmService
  {
    public static void Trigger(Alarm alarm)
    {
      if (CriteriaAreSatisfied(alarm))
      {
        AlarmService.TriggerAlarm(alarm.Nested1);
      }
    }

    private static bool IsSatisfied(TimeCriteria criterion)
    {
      switch (criterion)
      {
        case TimeCriteria.OnWeekend:
          return TimeService.IsWeekend();
        case TimeCriteria.OutsideWeekend:
          return TimeService.IsOutsideWeekend();
        case TimeCriteria.AtNight:
          return TimeService.IsNight();
        case TimeCriteria.DuringDay:
          return TimeService.IsDay();
        default:
          throw new ArgumentOutOfRangeException("criterion");
      }
    }

    public static void Disable(Alarm alarm)
    {
      AlarmService.DisableAlarm(alarm.Nested1);
    }

    public static void Dump(Alarm alarm)
    {
      Console.WriteLine("{ Timed Alarm active when: ");
      OutputCriteria(alarm);
      Console.WriteLine("When triggered : ");
      AlarmService.Dump(alarm.Nested1);
      Console.WriteLine(" }");
    }

    private static void OutputCriteria(Alarm alarm)
    {
      foreach (var criteria in alarm.TimeCriteria)
      {
        switch (criteria)
        {
          case TimeCriteria.OnWeekend:
            Console.WriteLine("it's weekend");
            break;
          case TimeCriteria.OutsideWeekend:
            Console.WriteLine("it's not weekend");
            break;
          case TimeCriteria.AtNight:
            Console.WriteLine("it's night");
            break;
          case TimeCriteria.DuringDay:
            Console.WriteLine("it's a day");
            break;
          default:
            throw new ArgumentOutOfRangeException();
        }
      }
    }

    private static bool CriteriaAreSatisfied(Alarm alarm)
    {
      return alarm.TimeCriteria.All(IsSatisfied);
    }

  }
}

File: Java/AlarmsProcedural/src/main/java/AlarmServices/TimedAlarmService.java
package AlarmServices;

import Adapters.TimeService;
import Enums.TimeCriteria;
import Main.Alarm;

import java.util.Arrays;


public class TimedAlarmService {
    public static void trigger(Alarm alarm) {
      if (Arrays.stream(alarm.timeCriterias).allMatch(c -> isSatisfied(c))) {
        AlarmService.triggerAlarm(alarm.nested1);
      }
    }

    private static boolean isSatisfied(TimeCriteria criteria) {
      switch (criteria) {
        case OnWeekend:
          return TimeService.isWeekend();
        case OutsideWeekend:
          return TimeService.isOutsideWeekend();
        case AtNight:
          return TimeService.isNight();
        case DuringDay:
          return TimeService.isDay();
        default:
          throw new IllegalArgumentException("criteria");
      }
    }

    public static void disable(Alarm alarm) {
      AlarmService.disableAlarm(alarm.nested1);
    }

    public static void dump(Alarm alarm) {
      System.out.println("{ Timed Alarm active when: ");
      outputCriterias(alarm);
      System.out.println("When triggered : ");
      AlarmService.dump(alarm.nested1);
      System.out.println(" }");
    }

    private static void outputCriterias(Alarm alarm) {
      for(TimeCriteria criteria : alarm.timeCriterias)
      {
        switch (criteria) {
          case OnWeekend:
            System.out.println("it's weekend");
            break;
          case OutsideWeekend:
            System.out.println("it's not weekend");
            break;
          case AtNight:
            System.out.println("it's night");
            break;
          case DuringDay:
            System.out.println("it's a day");
            break;
          default:
            throw new IllegalArgumentException("out or range");
        }
      }
    }
  }
